src/plugins/visualizations/public/utils/saved_visualize_utils.ts

```diff
@@ -0,0 +1,392 @@
+    .map((v) => v.appExtensions?.visualizations)
+    .filter(Boolean) as VisualizationsAppExtension[];
+  const extensionByType = extensions.reduce((acc, m) => {
+    return m!.docTypes.reduce((_acc, type) => {
+      acc[type] = m;
+      return acc;
+    }, acc);
+  }, {} as { [visType: string]: VisualizationsAppExtension });
+  const searchOption = (field: string, ...defaults: string[]) =>
+    _(extensions).map(field).concat(defaults).compact().flatten().uniq().value() as string[];
+  const searchOptions: SavedObjectsFindOptions = {
+    type: searchOption('docTypes', 'visualization'),
+    searchFields: searchOption('searchFields', 'title^3', 'description'),
+    search: search ? `${search}*` : undefined,
+    perPage: size,
+    page: 1,
+    defaultSearchOperator: 'AND' as 'AND',
+    hasReference: references,
+  };
+
+  const { total, savedObjects } = await savedObjectsClient.find<SavedObjectAttributes>(
+    searchOptions
+  );
+
+  return {
+    total,
+    hits: savedObjects.map((savedObject) => {
+      const config = extensionByType[savedObject.type];
+
+      if (config) {
+        return {
+          ...config.toListItem(savedObject),
+          references: savedObject.references,
+        };
+      } else {
+        return mapHitSource(visTypes, savedObject);
+      }
+    }),
+  };
+}
+
+export async function getSavedVisualization(
+  services: {
+    savedObjectsClient: SavedObjectsClientContract;
+    search: DataPublicPluginStart['search'];
+    dataViews: DataPublicPluginStart['dataViews'];
+    spaces: SpacesPluginStart;
+    savedObjectsTagging?: SavedObjectsTaggingApi;
+  },
+  opts?: any
+): Promise<VisSavedObject> {
+  if (typeof opts !== 'object') {
+    opts = { id: opts };
+  }
+
+  const id = (opts.id as string) || '';
+  const savedObject = {
+    id,
+    copyOnSave: false,
+    migrationVersion: opts.migrationVersion,
+    displayName: SAVED_VIS_TYPE,
+    getEsType: () => SAVED_VIS_TYPE,
+    getDisplayName: () => SAVED_VIS_TYPE,
+    searchSource: opts.searchSource ? services.search.searchSource.createEmpty() : undefined,
+  } as VisSavedObject & { _source: any };
+  const config = { injectReferences };
+
+  const defaultsProps = getDefaults(opts);
+
+  if (!id) {
+    Object.assign(savedObject, defaultsProps);
+    return savedObject;
+  }
+
+  const {
+    saved_object: resp,
+    outcome,
+    alias_target_id: aliasTargetId,
+  } = await services.savedObjectsClient.resolve<SavedObjectAttributes>(SAVED_VIS_TYPE, id);
+
+  if (!resp._version) {
+    throw new SavedObjectNotFound(SAVED_VIS_TYPE, id || '');
+  }
+
+  savedObject._source = _.cloneDeep(resp.attributes);
+
+  if (savedObject._source.visState) {
+    savedObject._source.visState = JSON.parse(savedObject._source.visState);
+  }
+
+  // assign the defaults to the response
+  _.defaults(savedObject._source, defaultsProps);
+
+  Object.assign(savedObject, savedObject._source);
+  savedObject.lastSavedTitle = savedObject.title;
```

Comment:
please check can we remove `lastSavedTitle`? Not sure that it's needed

Reply:
We use it here when check duplicates: 
![image](https://user-images.githubusercontent.com/16915480/135843333-d6bc07e2-e6b6-4f6b-8de5-35fc8657f55b.png)